torch/optim/adafactor.py

```diff
@@ -0,0 +1,311 @@
+                beta2=beta2,
+                d=group["d"],
+                lr=group["lr"],
+                weight_decay=group["weight_decay"],
+                eps1=eps1,
+                eps2=eps2,
+                grad_scale=getattr(self, "grad_scale", None),
+                found_inf=getattr(self, "found_inf", None),
+                has_complex=has_complex,
+                maximize=False,
+            )
+
+        return loss
+
+
+Adafactor.__doc__ = (
+    r"""Implements Adafactor algorithm.
+
+    The docs are currently wrong. Will fill in later.
+
+    For further details regarding the algorithm we refer to `Adafactor: Adaptive Learning Rates with Sublinear Memory Cost`_.
+    """
+    + r"""
+    Args:
+        params (iterable): iterable of parameters to optimize or dicts defining
+            parameter groups
+        lr (float, Tensor, optional): learning rate (default: 1e-3). A tensor LR
+            is not yet supported for all our implementations. Please use a float
+            LR if you are not also specifying fused=True or capturable=True.
+        betas (Tuple[float, float], optional): coefficients used for computing
+            running averages of gradient and its square (default: (0.9, 0.999))
+        eps (float, optional): term added to the denominator to improve
+            numerical stability (default: 1e-8)
+        weight_decay (float, optional): weight decay coefficient (default: 1e-2)
+
+    .. _Adafactor: Adaptive Learning Rates with Sublinear Memory Cost:
+        https://arxiv.org/pdf/1804.04235
+
+    """
+)
+
+
+def _single_tensor_adafactor(
+    params: List[Tensor],
+    grads: List[Tensor],
+    exp_avgs: List[Tensor],
+    row_vars: List[Tensor],
+    col_vars: List[Tensor],
+    state_steps: List[Tensor],
+    grad_scale: Optional[Tensor],
+    found_inf: Optional[Tensor],
+    *,
+    beta1: float,
+    beta2: float,
+    d: float,
+    lr: Union[Tensor, float],
+    weight_decay: float,
+    eps1: float,
+    eps2: float,
+    has_complex: bool,
+):
+    assert grad_scale is None and found_inf is None
+
+    if torch.jit.is_scripting():
+        # this assert is due to JIT being dumb and not realizing that the ops below
+        # have overloads to handle both float and Tensor lrs, so we just assert it's
+        # a float since most people using JIT are using floats
+        assert isinstance(lr, float)
+
+    for i, param in enumerate(params):
+        grad = grads[i]
+        if beta1 != 0:
+            exp_avg = exp_avgs[i]
+        step_t = state_steps[i]
+        row_var = row_vars[i]
+        col_var = col_vars[i]
+
+        if torch.is_complex(param):
+            grad = torch.view_as_real(grad)
+            if beta1 != 0:
+                exp_avg = torch.view_as_real(exp_avg)
+            param = torch.view_as_real(param)
+
+        # update step
+        step_t += 1
+
+        # Perform stepweight decay
+        param.mul_(1 - lr * weight_decay)
+
+        import math
```

Comment:
if `math` import is needed only for `sqrt(x)` - then maybe `x ** 0.5` can be used instead (same for `** 0.8`?) or is math.sqrt a bit faster?

Reply:
this is a draft! i don't plan on keeping this code lol

thanks for the suggestion

Reply:
there is also this :( https://stackoverflow.com/a/327011/445810 which says that `** 0.5` is a bit slower than math.sqrt. I hoped that `** 0.5` gets routed by python to math.sqrt under the hood...

Reply:
not in 3.10! 
<img width="505" alt="image" src="https://github.com/user-attachments/assets/23f6c5b7-5268-421c-b2ed-918397120c02">

Reply:
Also, for the same reasons dispatch_sqry handler can probably be eliminated in other optimizers